src/AutoMapper/TypePair.cs

```diff
@@ -32,7 +30,7 @@ public override bool Equals(object obj)
             return obj is MapRequest && Equals((MapRequest) obj);
         }
 
-        public override int GetHashCode() => _hashcode;
+        public override int GetHashCode() => HashCodeCombiner.Combine(RequestedTypes, RuntimeTypes);
```

Comment:
```
private int? _hashcode;
public override int GetHashCode() => _hashcode ?? (_hashcode = HashCodeCombiner.Combine(RequestedTypes, RuntimeTypes)).Value;
```
Every so slightly optimized by caching like original code.
Same with other implementations and each map does a couple of comparisons.
And inline collections will have to compare again as well.
Might be significant.